packages/block-editor/src/components/block-list/block.js

```diff
@@ -540,9 +543,39 @@ export default compose(
 	pure,
 	applyWithSelect,
 	applyWithDispatch,
+	( WrappedComponent ) => ( props ) => {
+		const [ subscriptions, setSubscriptions ] = useState( new Set() );
+		const registry = useRegistry();
+		const attributes = useSelect(
+			( select ) =>
+				Object.fromEntries(
+					Object.entries(
+						select( blockEditorStore ).getBlockAttributes(
+							props.clientId
+						)
+					).filter( ( [ key ] ) => subscriptions.has( key ) )
+				),
+			[ subscriptions, props.clientId ]
+		);
+
+		const proxy = new Proxy( attributes, {
+			get( target, name ) {
+				if ( ! subscriptions.has( name ) ) {
+					window.queueMicrotask( () => {
+						setSubscriptions( ( subs ) => subs.add( name ) );
+					} );
+					return registry
+						.select( blockEditorStore )
+						.getBlockAttributes( props.clientId )[ name ];
+				}
+				return target[ name ];
+			},
+		} );
+		return <WrappedComponent { ...props } attributes={ proxy } />;
```

Comment:
I think that `subscriptions` don't need to be state, and adding a subscription doesn't need to trigger a state update and a rerender. It can be a ref:
```js
const subscriptions = useRef( new Set() );
```
And the `useSelect` doesn't need to have `subscriptions` as dependency.

The workflow is like this:
1. Initially, when we didn't read any property of the attributes proxy yet, `subscriptions` is an empty set and the `useSelect` will always return `{}`. No store updates will ever trigger a rerender of `WrappedComponent` because `useSelect` continues to return `{}`.
2. Component reads `attributes.content`. We get the value directly from `registry.select`, and also call `subscriptions.current.add( 'content' )`. Synchronously, no state, no microtask.
3. On next store update, the `useSelect` callback will be called, and this time `subscriptions is `{ 'content' }` and `useSelect` will return `content`. The return value has changes and will trigger a rerender.

One remaining problem is that with repeated calls on the proxy:
```js
attributes.content;
attributes.content;
```
The second call will see a new value of `subscriptions`, which already includes `content`, and will return `undefined`, because `target[ name ]` is not present. Instead of reading `subscriptions` from the ref in closure, we'll need to pass a copy:
```js
const proxy = createProxy( attributes, cloneSet( subscriptions.current ) );
```
That way the `attributes` and `subscriptions` objects never get out of sync.

Or, maybe the proxy doesn't need to _read_ `subscriptions` at all? It can simply do:
```js
if ( target.hasOwnProperty( name ) ) {
  return target[ name ];
}

subscriptions.current.add( name );
return registry.select().getAttributes()[ name ];
```

Reply:
> The second call will see a new value of subscriptions, which already includes content, and will return `undefined`

I see this behavior also with the current version, with `subscriptions` as state. Try to render a Heading block and log the `level` attribute on each render. You'll see a render where it's `undefined`, and a React warning that you're trying to render an unknown HTML tag `<hundefined>` (`'h' + level`).

This happens because the `setSubscriptions( s => s.add( name ) )` callback also modifies the _original_ `subscriptions` set, before actually executing the state update and selecting updated `attributes`.

Reply:
If you don't mind I pushed a commit that fixes the subscriptions update. It fixes incorrect rendering, maybe it will affect the performance results, too?

Reply:
Interesting, thank you!